### PR TITLE
[feat] add split_dim arg to reversible, remove retain_grad, add benchmark_reversible

### PR DESCRIPTION
**Pull request by ClashLuke**
This PR removes the repeated chunk and cat operations in xformers' RevNet code. This way, the RevNet implementation will become a little bit faster.\
I'd strongly recommend calling a library like [MemCNN](https://github.com/silvandeleemput/memcnn) or [RevLib](https://github.com/ClashLuke/revlib) directly as they make it easier to switch the coupling function and generally give the user more freedom.

Unfortunately, I can't sign the CLA at the moment, as it keeps saying
> Sorry, something went wrong.
> We're working on getting this fixed as soon as we can.

### CODE REVIEW

**Review by dianaml0 — CHANGES_REQUESTED** [review]
Thanks a lot @ClashLuke for your contribution and for testing out the impact of these changes to reversible layers! The benchmark is a really great addition! It seems like removing the `cat` and `chunk` operations didn't have the expected speed up for some reason - perhaps we can merge in the benchmarking code but leave `reversible.py` as it was for the sake of simplicity?

File: xformers/components/reversible.py [hunk 1]
```diff
@@ -135,4 +133,6 @@ def forward(self, x, arg_route=(True, False), **kwargs):
         f_args, g_args = map(lambda route: kwargs if route else {}, arg_route)
         block_kwargs = {"f_args": f_args, "g_args": g_args}
 
-        return _ReversibleFunction.apply(x, self.blocks, block_kwargs)
+        x1, x2 = x.chunk(2, 2)
```
[the remark this answers is not in the document]
Comment (ClashLuke) [reply xformers/components/reversible.py]: Following MemCNN's and RevLib's API, I just made it a parameter with the default set to -1.

**Comment by ClashLuke** [comment]
I just gave it a try on a rented A100, and it seems a little faster with the current setup, and it seems like the speedup is negligible, especially for large batch sizes. If you find the original code simpler, let's revert.


<details>
<summary>Batch * Sequence in (16ki, 2ki, 128) </summary>

```
Upstream:

 --- Type: torch.float16 --- 
| Units: runtime in ms, lower is better          |Batch=16384, Features=32, Depth=4|Batch=16384, Features=32, Depth=32|Batch=16384, Features=32, Depth=256|Batch=2048, Features=256, Depth=4|Batch=2048, Features=256, Depth=32|Batch=2048, Features=256, Depth=256|Batch=128, Features=4096, Depth=4|Batch=128, Features=4096, Depth=32|Batch=128, Features=4096, Depth=256|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|residual - fw                                   |0.17                |2.61                |33.81               |0.16                |4.15                |30.77               |0.42                |3.25                |32.05               |
|reversible - fw                                 |0.27                |6.52                |56.69               |0.39                |5.55                |52.31               |0.44                |6.79                |56.95               |


 --- Type: torch.float32 --- 
| Units: runtime in ms, lower is better          |Batch=16384, Features=32, Depth=4|Batch=16384, Features=32, Depth=32|Batch=16384, Features=32, Depth=256|Batch=2048, Features=256, Depth=4|Batch=2048, Features=256, Depth=32|Batch=2048, Features=256, Depth=256|Batch=128, Features=4096, Depth=4|Batch=128, Features=4096, Depth=32|Batch=128, Features=4096, Depth=256|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|residual - fw                                   |0.17                |2.41                |29.93               |0.18                |2.51                |28.32               |0.76                |6.01                |47.60               |
|reversible - fw                                 |0.18                |5.97                |42.80               |0.19                |6.34                |41.91               |0.78                |6.08                |47.86               |


 --- Type: torch.float16 --- 
| Units: runtime in ms, lower is better          |Batch=16384, Features=32, Depth=4|Batch=16384, Features=32, Depth=32|Batch=16384, Features=32, Depth=256|Batch=2048, Features=256, Depth=4|Batch=2048, Features=256, Depth=32|Batch=2048, Features=256, Depth=256|Batch=128, Features=4096, Depth=4|Batch=128, Features=4096, Depth=32|Batch=128, Features=4096, Depth=256|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|residual - fw+bw                                |1.47                |10.97               |115.48              |0.94                |14.53               |194.51              |2.16                |14.96               |119.33              |
|reversible - fw+bw                              |3.91                |31.17               |225.91              |3.06                |35.75               |241.17              |3.47                |32.27               |217.62              |


 --- Type: torch.float32 --- 
| Units: runtime in ms, lower is better          |Batch=16384, Features=32, Depth=4|Batch=16384, Features=32, Depth=32|Batch=16384, Features=32, Depth=256|Batch=2048, Features=256, Depth=4|Batch=2048, Features=256, Depth=32|Batch=2048, Features=256, Depth=256|Batch=128, Features=4096, Depth=4|Batch=128, Features=4096, Depth=32|Batch=128, Features=4096, Depth=256|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|residual - fw+bw                                |1.31                |11.86               |78.94               |0.76                |13.88               |87.93               |3.71                |26.66               |212.21              |
|reversible - fw+bw                              |3.60                |28.95               |230.99              |3.11                |24.63               |216.43              |4.22                |33.24               |264.72              |

========================

Proposed:

 --- Type: torch.float16 --- 
| Units: runtime in ms, lower is better          |Batch=16384, Features=32, Depth=4|Batch=16384, Features=32, Depth=32|Batch=16384, Features=32, Depth=256|Batch=2048, Features=256, Depth=4|Batch=2048, Features=256, Depth=32|Batch=2048, Features=256, Depth=256|Batch=128, Features=4096, Depth=4|Batch=128, Features=4096, Depth=32|Batch=128, Features=4096, Depth=256|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|residual - fw                                   |0.17                |3.62                |26.28               |0.16                |3.91                |33.23               |0.42                |3.74                |31.64               |
|reversible - fw                                 |0.15                |4.90                |44.41               |0.37                |7.11                |49.51               |0.44                |5.00                |44.59               |


 --- Type: torch.float32 --- 
| Units: runtime in ms, lower is better          |Batch=16384, Features=32, Depth=4|Batch=16384, Features=32, Depth=32|Batch=16384, Features=32, Depth=256|Batch=2048, Features=256, Depth=4|Batch=2048, Features=256, Depth=32|Batch=2048, Features=256, Depth=256|Batch=128, Features=4096, Depth=4|Batch=128, Features=4096, Depth=32|Batch=128, Features=4096, Depth=256|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|residual - fw                                   |0.16                |3.36                |34.08               |0.17                |3.42                |27.66               |0.76                |6.00                |47.43               |
|reversible - fw                                 |0.18                |4.47                |48.93               |0.19                |6.34                |43.61               |0.78                |6.76                |47.71               |


 --- Type: torch.float16 --- 
| Units: runtime in ms, lower is better          |Batch=16384, Features=32, Depth=4|Batch=16384, Features=32, Depth=32|Batch=16384, Features=32, Depth=256|Batch=2048, Features=256, Depth=4|Batch=2048, Features=256, Depth=32|Batch=2048, Features=256, Depth=256|Batch=128, Features=4096, Depth=4|Batch=128, Features=4096, Depth=32|Batch=128, Features=4096, Depth=256|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|residual - fw+bw                                |1.38                |10.50               |89.10               |0.69                |10.65               |100.32              |2.15                |15.35               |118.91              |
|reversible - fw+bw                              |3.54                |30.13               |209.89              |4.43                |31.94               |238.64              |3.86                |33.39               |263.07              |


 --- Type: torch.float32 --- 
| Units: runtime in ms, lower is better          |Batch=16384, Features=32, Depth=4|Batch=16384, Features=32, Depth=32|Batch=16384, Features=32, Depth=256|Batch=2048, Features=256, Depth=4|Batch=2048, Features=256, Depth=32|Batch=2048, Features=256, Depth=256|Batch=128, Features=4096, Depth=4|Batch=128, Features=4096, Depth=32|Batch=128, Features=4096, Depth=256|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|residual - fw+bw                                |1.26                |11.99               |103.16              |1.21                |12.06               |84.13               |3.59                |26.60               |212.19              |
|reversible - fw+bw                              |3.62                |28.10               |189.94              |3.61                |29.38               |201.67              |4.49                |33.17               |264.14              |
```

</details>

<details>
<summary>Batch * Sequence in (512ki, 64ki, 4ki) </summary>

```
Upstream:

 --- Type: torch.float16 ---
| Units: runtime in ms, lower is better          |Batch=524288, Features=32, Depth=4|Batch=524288, Features=32, Depth=32|Batch=524288, Features=32, Depth=256|Batch=65536, Features=256, Depth=4|Batch=65536, Features=256, Depth=32|Batch=65536, Features=256, Depth=256|Batch=4096, Features=4096, Depth=4|Batch=4096, Features=4096, Depth=32|Batch=4096, Features=4096, Depth=256|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|residual - fw                                   |1.85                |13.82               |110.34              |1.78                |14.24               |113.61              |5.65                |45.17               |347.51              |
|reversible - fw                                 |1.94                |14.03               |110.63              |1.99                |14.47               |113.81              |5.92                |45.78               |348.78              |


 --- Type: torch.float32 ---
| Units: runtime in ms, lower is better          |Batch=524288, Features=32, Depth=4|Batch=524288, Features=32, Depth=32|Batch=524288, Features=32, Depth=256|Batch=65536, Features=256, Depth=4|Batch=65536, Features=256, Depth=32|Batch=65536, Features=256, Depth=256|Batch=4096, Features=4096, Depth=4|Batch=4096, Features=4096, Depth=32|Batch=4096, Features=4096, Depth=256|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|residual - fw                                   |3.08                |24.62               |196.90              |3.06                |24.54               |196.16              |10.57               |85.51               |671.90              |
|reversible - fw                                 |3.41                |24.96               |197.29              |3.40                |24.87               |196.55              |10.91               |85.82               |677.07              |


 --- Type: torch.float16 ---
| Units: runtime in ms, lower is better          |Batch=524288, Features=32, Depth=4|Batch=524288, Features=32, Depth=32|Batch=524288, Features=32, Depth=256|Batch=65536, Features=256, Depth=4|Batch=65536, Features=256, Depth=32|Batch=65536, Features=256, Depth=256|Batch=4096, Features=4096, Depth=4|Batch=4096, Features=4096, Depth=32|Batch=4096, Features=4096, Depth=256|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|residual - fw+bw                                |4.08                |31.18               |247.70              |4.42                |33.99               |270.20              |16.37               |123.36              |965.07              |
|reversible - fw+bw                              |6.37                |45.73               |360.55              |6.72                |48.81               |385.56              |22.70               |167.17              |1306.73             |


 --- Type: torch.float32 ---
| Units: runtime in ms, lower is better          |Batch=524288, Features=32, Depth=4|Batch=524288, Features=32, Depth=32|Batch=524288, Features=32, Depth=256|Batch=65536, Features=256, Depth=4|Batch=65536, Features=256, Depth=32|Batch=65536, Features=256, Depth=256|Batch=4096, Features=4096, Depth=4|Batch=4096, Features=4096, Depth=32|Batch=4096, Features=4096, Depth=256|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|residual - fw+bw                                |7.33                |56.22               |448.13              |7.74                |59.61               |474.78              |31.30               |255.06              |1924.65             |
|reversible - fw+bw                              |11.27               |81.70               |646.12              |11.68               |85.28               |673.13              |42.79               |338.19              |2561.36             |


=====================

Proposed:

 --- Type: torch.float16 ---
| Units: runtime in ms, lower is better          |Batch=524288, Features=32, Depth=4|Batch=524288, Features=32, Depth=32|Batch=524288, Features=32, Depth=256|Batch=65536, Features=256, Depth=4|Batch=65536, Features=256, Depth=32|Batch=65536, Features=256, Depth=256|Batch=4096, Features=4096, Depth=4|Batch=4096, Features=4096, Depth=32|Batch=4096, Features=4096, Depth=256|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|residual - fw                                   |1.97                |13.82               |110.45              |1.78                |14.24               |113.69              |5.65                |44.96               |347.49              |
|reversible - fw                                 |1.94                |14.04               |110.56              |1.99                |14.46               |114.05              |5.94                |45.98               |350.02              |


 --- Type: torch.float32 ---
| Units: runtime in ms, lower is better          |Batch=524288, Features=32, Depth=4|Batch=524288, Features=32, Depth=32|Batch=524288, Features=32, Depth=256|Batch=65536, Features=256, Depth=4|Batch=65536, Features=256, Depth=32|Batch=65536, Features=256, Depth=256|Batch=4096, Features=4096, Depth=4|Batch=4096, Features=4096, Depth=32|Batch=4096, Features=4096, Depth=256|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|residual - fw                                   |3.08                |24.63               |197.04              |3.07                |24.56               |196.38              |10.56               |85.49               |672.41              |
|reversible - fw                                 |3.41                |24.97               |197.30              |3.40                |24.88               |196.73              |10.91               |85.81               |677.29              |


 --- Type: torch.float16 ---
| Units: runtime in ms, lower is better          |Batch=524288, Features=32, Depth=4|Batch=524288, Features=32, Depth=32|Batch=524288, Features=32, Depth=256|Batch=65536, Features=256, Depth=4|Batch=65536, Features=256, Depth=32|Batch=65536, Features=256, Depth=256|Batch=4096, Features=4096, Depth=4|Batch=4096, Features=4096, Depth=32|Batch=4096, Features=4096, Depth=256|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|residual - fw+bw                                |4.08                |31.24               |247.76              |4.42                |33.96               |270.01              |16.33               |123.62              |965.01              |
|reversible - fw+bw                              |6.37                |45.68               |360.07              |6.73                |48.79               |385.51              |22.82               |168.32              |1307.25             |


 --- Type: torch.float32 --- 
| Units: runtime in ms, lower is better          |Batch=524288, Features=32, Depth=4|Batch=524288, Features=32, Depth=32|Batch=524288, Features=32, Depth=256|Batch=65536, Features=256, Depth=4|Batch=65536, Features=256, Depth=32|Batch=65536, Features=256, Depth=256|Batch=4096, Features=4096, Depth=4|Batch=4096, Features=4096, Depth=32|Batch=4096, Features=4096, Depth=256|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|residual - fw+bw                                |7.34                |56.26               |448.10              |7.75                |59.73               |475.17              |31.35               |254.32              |1925.63             |
|reversible - fw+bw                              |11.29               |81.75               |646.46              |11.68               |85.38               |673.95              |42.83               |339.35              |2561.78             |
```

**Comment by blefaudeux** [comment]
@ClashLuke thanks for the changes and sorry for the broken test, should be fixed now if you rebase onto main :/ really sorry about the experience

**Review by blefaudeux — APPROVED** [review]
Looks great to me, thanks again @ClashLuke  !

**Review by dianaml0 — APPROVED** [review]
Looks great!

**Pull request merged**